#44 Refactor sitemap registration pattern.

joemcgill opened the pull request:
### Issue Number
This fixes #44.

### Description
This makes the Core_Sitemaps_Registry a property of the Core_Sitemaps class and moves the responsibility for registering sitemaps from each sitemap provider to the main Core_Sitemaps class. This allows each provider to expose the properties required for setting up and rendering sitemaps without needing to have any knowledge of the registry at all so we're not passing around as many objects between parts of the system.

### Type of change
Please select the relevant options:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### Steps to test
- [ ] Make sure to flush rewrite rules after checking out this branch.
- [ ] Visit /sitemap.xml to see the sitemap index including links to the two sitemaps for pages and posts (currently with incorrect URLs).
- [ ] Visit /sitemap-pages.xml to see the sitemap for pages.
- [ ] Visit /sitemap-posts.xml to see the sitemap for posts.

### Acceptance criteria
- [x] My code follows WordPress coding standards.
- [x] I have performed a self-review of my own code.
- [ ] If the changes are visual, I have cross browser / device tested.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [ ] My changes generate no new warnings.
- [x] I have added test instructions that prove my fix is effective or that my feature works.

Commit:
Refactor sitemap registration pattern.

This makes the Core_Sitemaps_Registry a property of the Core_Sitemaps class and moves the responsibility for registering sitemaps from each sitemap provider to the main Core_Sitemaps class. This allows each provider to expose the properties required for setting up and rendering sitemaps without needing to have any knowledge of the registry at all so we're not passing around as many objects between parts of the system.

joemcgill commented on inc/class-sitemaps-index.php:
```diff
@@ -10,7 +10,7 @@
  * Builds the sitemap index page that lists the links to all of the sitemaps.
  */
-class Core_Sitemaps_Index extends Core_Sitemaps_Provider {
+class Core_Sitemaps_Index {
```
I've specifically made this change because I think we need to maintain a distinction between the sitemap index and the individual sitemaps which will be listed in the index. There are definitely common patterns between them, but I wanted to decouple these from a conceptual point of view so our code can be more clear and keep a separation of concerns between the code that provides and index, and the code that provides a sitemap.

joemcgill commented on inc/class-sitemaps-index.php:
```diff
@@ -22,22 +22,18 @@ class Core_Sitemaps_Index extends Core_Sitemaps_Provider {
-	 * @uses add_action()
-	 * @uses add_filter()
 	 */
 	public function bootstrap() {
```
Should we change this method name to `setup_sitemap()` to be more clear?

svandragt replied:
I'd definately be in favour of moving towards a more explicit description of what the bootstrap function does. as this avoids putting code that should be in a different method structurally.

Commit:
Clean up docblocks.

Commit:
Update docs and variable names for 'core_sitemaps_get_sitemaps()'

The docs and variable names should accurately reflect that this function is returning sitemap providers, not sitemap urls.

joemcgill commented on inc/functions.php:
```diff
@@ -0,0 +1,19 @@
+ *
+ * @return array $sitemaps A list of registered sitemap providers.
+ */
+function core_sitemaps_get_sitemaps() {
```
This is the first of many public functions we'll want to create that returns data from the global `Core_Sitemaps` object so that people using sitemaps functionality won't have to pass objects around when working with sitemaps data.

svandragt commented on inc/class-sitemaps-pages.php:
```diff
@@ -11,28 +11,33 @@ class Core_Sitemaps_Pages extends Core_Sitemaps_Provider {
-	public function register_sitemap() {
-		$this->registry->add_sitemap( $this->name, '^sitemap-pages\.xml$', esc_url( $this->get_sitemap_url( $this->name ) ) );
-	}
+	public $slug = 'page';
```
```suggestion
	public $slug = 'pages';
```
Does not match $route, and the post provider is using plural here.

svandragt commented on inc/class-sitemaps-registry.php:
```diff
@@ -76,16 +64,28 @@ public function get_sitemaps() {
+			if ( ! $wp_rewrite->using_permalinks() ) {
+				$url = add_query_arg( 'sitemap', $name, home_url( '/' ) );
+			}
 		}
```
This code can be simplified to:
```php
$home_url_append = '';
if ( 'index' !== $name ) {
	$home_url_append = '-' . $name;
}
$url = home_url( sprintf( '/sitemap%1$s.xml', $home_url_append ) );

if ( ! $wp_rewrite->using_permalinks() ) {
	$url = add_query_arg( 'sitemap', $name, home_url( '/' ) );
}
```

svandragt reviewed (changes requested):
Nice work moving forward the plugin architecture.

svandragt commented on inc/class-sitemaps-renderer.php:
```diff
@@ -14,13 +14,13 @@ class Core_Sitemaps_Renderer {
 		foreach ( $sitemaps as $link ) {
 			$sitemap = $sitemap_index->addChild( 'sitemap' );
-			$sitemap->addChild( 'loc', esc_url( $link['slug'] ) );
+			$sitemap->addChild( 'loc', esc_url( $link->slug ) );
```
```xml
<sitemapindex>
<sitemap>
<loc>http://posts</loc>
<lastmod>2004-10-01T18:23:17+00:00</lastmod>
</sitemap>
<sitemap>
<loc>http://page</loc>
<lastmod>2004-10-01T18:23:17+00:00</lastmod>
</sitemap>
</sitemapindex>
```
These links need to use get_sitemap_url()

svandragt reviewed (changes requested):
Found an issue with the index sitemap links

kirstyburgoine committed:
44: rename bootstrap() to setup_sitemap() in index

kirstyburgoine committed:
44: Update $slug for pages

kirstyburgoine committed:
44: Move get_sitemap_url to renderer

This is because this is only needed in renderer and index, if we leave it registery then the registry needs to be instaiated in these two classes as well

kirstyburgoine commented:
@googlebot I consent

kirstyburgoine committed:
44: update get_sitemap_url() for robots.txt

kirstyburgoine committed:
44: Add __construct() to Core_Sitemaps_Index

svandragt commented on inc/class-sitemaps-renderer.php:
```diff
@@ -9,18 +9,39 @@
 		foreach ( $sitemaps as $link ) {
 			$sitemap = $sitemap_index->addChild( 'sitemap' );
-			$sitemap->addChild( 'loc', esc_url( $link['slug'] ) );
+			$sitemap->addChild( 'loc', esc_url( $this->get_sitemap_url( $link->slug ) ) );
```
get_sitemap_url() expects the name not the slug, although this code happens to work because they're the same.
```suggestion
			$sitemap->addChild( 'loc', esc_url( $this->get_sitemap_url( $link->name ) ) );
```

svandragt reviewed (changes requested):
One small thing I will push up the change.
Outwith this PR what would you think about adding get_sitemap_url() a function in inc/functions.php looking at it again it seems that might fit even better.

svandragt committed:
get_sitemap_url() expects name.

svandragt commented:
@googlebot I consent.
